### DISCUSSION TITLE
Refactoring Magnification Calculations

### DISCUSSION THREAD

**Discussion by jenniferyee** [rpoleski/MulensModel#49] [Ideas]
@rpoleski @alpv95 : (C) is the real solution, but I include my intermediate thinking (B), so you can see why it doesn't work.

Use MagnificationCurve to link Data to model calculations:

A) Currently, data magnification is calculated from

  1. Model.get_magnification(data.times, **magnification_kwargs)

  2. which passes loads of stuff to MagnificationCurve

  3. which then creates a Trajectory object and proceeds to calculation the
  magnification.

B) Instead, data_magnification could be calculated by

  1. create a Trajectory object from the data and the model parameters

  2. pass the Trajectory to MagnificationCurve

  3. calculate the magnification from MagnificationCurve.get_magnification()

There are two advantages to this approach:

1. It eliminates most of the **magnification_kwargs (except gamma/bandpass) by
making them part of Trajectory.

2. It creates an object (MagnificationCurve) that can preserve all kinds of
information that can be used to calculate derivatives.

C) One problem with the (B) approach is there is no clear way to include binary
source models. This could be solved by

   1. create a MagnificationCurve object from the data and the model parameters

   2. the MagnificationCurve then creates the Trajectory object(s).

   3. calculate the magnification from MagnificationCurve.get_magnification()

This retains the advantages of (B) while also making it straightforward to
handle both gamma/bandpass information AND binary sources.

So then, FitData._calculate_magnifications() becomes

```python
def _calculate_magnifications(self, bad=True):
    if self.dataset.badpass is not None:
        gamma = self.model.get_limb_coeff_gamma(self.dataset.bandpass)
    else:
        gamma = None    
    
    self.mag_curve = MagnificationCurve(
        self.dataset.times, bad=self.dataset.bad, 
        satellite_skycoord=self.dataset.satellite_skycoord, 
        gamma=gamma)
    
    self._data_magnification = self.mag_curve.get_magnification(bad=bad)
 ```

And, the following properties could be preserved: 
    self.mag_curve.trajectory
    self.mag_curve.a_pspl
    self.mag_curve.B_0, B_1, B_0_prime, B_1_prime
    etc.

**Comment by rpoleski** [comment]
I suggest somebody (e.g., @jenniferyee) implements this idea in a new branch and then we can think about it in detail. In general, I like the idea.

Minor comment: `self.mag_curve` -> `self._mag_curve` :)
